주식 구매 기능 구현

hek316 opened the pull request:
## 작업 개요
사용자가 특정 주식을 구매할 수 있는 API 기능을 구현했습니다.

## 작업 상세
- 주식 주문 엔티티 (`StockOrderEntity`, `OrderType`, `OrderStatus`) 설계
- `POST /api/sorders` 엔드포인트 추가
- 멱등성 검사 (`IdempotencyValidator`)로 중복 주문 방지
- 주문 요청 시 KIS API 연동 및 구매 이력 저장

## TODO
- [ ] 주문 완료 여부를 확인하는 스케줄러 구현
Close #45

hek316 committed:
feat: StockOrderEntity, AccountEntity 1차 설계

hek316 committed:
feat(order): 주식 주문 로직 구현
- 주문 처리에 대한 멱등성 검사를 위한 IdempotencyValidator 적용
- Member, 종목 코드 유효성 검증 로직을 각각 MemberValidator, StockCodeValidator로 분리
- KIS 주문 요청 시 네트워크 타임아웃(WebClientRequestException) 발생 시 1회 재시도 처리 추가
- 주문 요청 처리 후 OrderResult 응답 반환

hek316 committed:
Refactor: AccountEntity 변수명 수정으로 가독성 개선
- AccountEntity의 변수명을 의미에 맞게 변경하여 코드 가독성 향상

외부 API 호출 순서를 마지막으로 조정하고, 상태 체크 로직 개선
- StockOrder 엔티티 저장 후 KIS 네트워크 호출 순서 변경
- 주문 상태 체크를 OrderStatus 컬럼을 통해 수행하도록 로직 개선

f-lab-cacio commented on src/main/java/com/flab/mars/domain/vo/OrderStockVO.java:
```diff
@@ -0,0 +1,54 @@
+@ToString
+@NoArgsConstructor
+@AllArgsConstructor
+public class OrderStockVO {
```
도메인을 표현하기보단 request dto 를 그대로 투영하고 있다는 느낌이 더 강하네요.
매수/매도, 지정가주문/시장가 주문에 대한 도메인 객체를 분리할 수 있지 않을까요?
또 인증 관련 필드들도 OrderStock 이라는 도메인 객체가 아닌 다른 객체로 표현하는건 어떨까요?

hek316 replied:
피드백 주신 내용 반영하여, 매수/매도, 지정가주문/시장가 주문 도메인 객체로 분리해 단일 책임 원칙을 지키고, 객체간 책임을 명확히 하여 유지보수성과 가독성을 높였습니다!

f-lab-cacio commented on src/main/java/com/flab/mars/domain/service/StockOrderService.java:
```diff
@@ -0,0 +1,98 @@
+            kisClientOrder.orderStock(kisOrderStockVO);
+        } catch (WebClientRequestException e) {
+            if (isTimeoutException(e)) {
+                kisClientOrder.orderStock(kisOrderStockVO);
```
timeout 으로 실패한 경우는 kis 쪽에서 요청을 제대로 받고 아직 처리중인건지, 요청을 받지 못한 것인지 알 수 없는데 orderStock 을 다시 호출해도 문제 없을까요?
만약 전자인 경우였다면 어떻게 될까요?

hek316 replied:
중복 주문이 발생할 위험이 있습니다.  왜냐하면 timeout 은 요청이 도달하지 않았을 수 있고,  도달했지만 응답만 늦을 것일 수 도 있기 때문입니다. 

타임아웃 시 주식일별주문체결조회 API 를 이용해 주문 확인로직을 추가해서 처리해보겠습니다!

f-lab-cacio commented on src/main/java/com/flab/mars/domain/service/StockOrderService.java:
```diff
@@ -0,0 +1,98 @@
+        }
+
+        orderEntity.markPending();
+        stockOrderRepository.save(orderEntity);
```
jpa 에서는 엔티티 변경에 대해선 dirty checking 을 통해 update 하는 것이 권장되는 패턴입니다.
(https://vladmihalcea.com/best-spring-data-jparepository/)

hek316 replied:
이미 영속상태 엔티티는 필드 값만 수정해도 `dirty checking`으로 자동으로 `UPDATE` 쿼리가 발생하니, 이 방식이 가장 효율적이네요.

`save()`는 새로운 엔티티를 저장 용도로만 사용하고,
이미 영속 상태인 엔티티는 필드만 수정한 뒤 트랜잭션을 커밋해 update 되도록 해야 불필요한 `merge()` 호출과 그로 인해 `select`, `cascade` 병합 오버헤드를 피할 수 있겠네요!

f-lab-cacio commented on src/main/java/com/flab/mars/db/entity/StockOrderEntity.java:
```diff
@@ -0,0 +1,67 @@
+    @Column(name = "idempotency_key", nullable = false, unique = true, length = 100)
+    private String idempotencyKey;
+
+    // == 비즈니스로직 ==
+    //  도메인의 책임을 객체에 집중시켜 코드의 일관성관 유지 보수정을 높힘.
+    //  도메인의 중요한 상태전이나 규칙이 있을때, set 메서드보다 의미있는 동작을 표현하고 싶을때
+
+    public void markPending() {
+        this.orderStatus = OrderStatus.PENDING;
+    }
+
+    public void markFilled() {
+        this.orderStatus = OrderStatus.FILLED;
+    }
+
+    public void markCanceled() {
+        this.orderStatus = OrderStatus.CANCELED;
+    }
```
👍

f-lab-cacio commented on src/main/java/com/flab/mars/db/entity/AccountEntity.java:
```diff
@@ -0,0 +1,25 @@
+    @JoinColumn(name = "member_id")
+    private MemberEntity member;
+
+    private String accountPrefix;  // 계좌번호 앞 8자리
+    private String accountSuffix;  // 계좌번호 뒤 2자리
+
+    private boolean defaultFlag;
```
이 필드들은 어떤 목적으로  있는걸까요?

hek316 replied:
한투 API를 통해 주식을 구매할 때, 계좌번호 앞 8자리(accountPrefix)와 뒤 2자리(accountSuffix)가 필수 파라미터로 요구됩니다.
또한, 한 명의 사용자가 여러 계좌를 보유할 수 있기 때문에, defaultFlag를 통해 기본 계좌 여부를 명시하고자 합니다.

 accountPrefix와 accountSuffix는 항상 함께 사용되므로, 의미상 하나의 값 객체로 묶어 Embedded 타입으로 분리할까요~?

f-lab-cacio commented on src/main/java/com/flab/mars/config/RootConfig.java:
```diff
@@ -18,7 +18,7 @@ public class RootConfig {
     @Bean
     public WebClient webClient(WebClient.Builder builder) {
-        return builder.baseUrl(KISApiUrls.BASE_URL) // KIS API의 기본 URL 설정
+        return builder.baseUrl(KISApiUrls.MOCK_BASE_URL) // KIS API의 기본 URL 설정
```
별도 PR 에서 config 객체를 정의해서 yml 에 phase 별로 다른 url 값을 주입하도록 구성해보시죠~

hek316 replied:
[넵!](https://github.com/f-lab-edu/mars/pull/55)

hek316 committed:
refactor: 주문 도메인 객체 분리 및 인증 정보 분리

- OrderStockVO가 요청 DTO 구조를 그대로 반영하고 있어 도메인 역할을 제대로 수행하지 못하는 문제 개선
- 매수/매도, 지정가/시장가 주문을 도메인 객체로 각각 분리하여 의도를 명확히 표현
- 인증 관련 필드(appKey, appSecret, accessToken)는 별도의 AuthInfo 객체로 분리하여 책임 분리
- 값 객체를 불변(immutable) 구조로 정의하여 유지보수성과 디버깅 용이성 향상

hek316 committed:
refactor: 서로 배타적으로 초기화되는 인스턴스 변수를 기준으로 시장가 주문과 지정가 주문을 분리하여 단일 책임 원칙 준수

- PriceStrategy 인터페이스 도입으로 가격 전략을 분기 처리하도록 개선
- OrderVo 클래스 내에서 주문 공통 정보와 가격 전략을 결합하여 책임 분리 및 확장성 강화

hek316 committed:
refactor: JPA 더티 체키 활용을 위한 @Transactional 적용 및 중복 save 제거

- 주문 처리 서비스에 @Transactional 추가하여 트랜잭션 내 더티 체킹 적용
- 주문 상태 변경 시 불필요한 save() 호출 제거 (자동 update 반영)

hek316 committed:
refactor: 주문 요청 timeout 대응을 위한 TODO 추가

- timeout일 경우 재요청 전에 중복 주문 여부 확인 필요 (TODO로 명시)
- 추후 KIS 주식일별주문체결조회 API를 통해 중복 방지 로직 추가 예정

hek316 committed:
refactor: 오타 수정
